### PR TITLE
introduce TensorBase::mutable_data_ptr<T>

### PR DESCRIPTION
**Pull request by dagitses**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98781
* #98780
* #98756
* #98755
* #98754
* #98753
* #98752
* #98751
* #98750
* #98734
* #98162
* #98163
* __->__ #97874

See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr<T>() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44497685](https://our.internmc.facebook.com/intern/diff/D44497685/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44497685/)!

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -1554,29 +1554,41 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
...
+      return (self.*data_ptr_impl_accessor)();
     }
   };
 
  public:
+  /**
+   * More efficient helper for Tensor::data_ptr(). Like data<T>(), but
+   * does not do a type check. Unlike the untemplated data(), does
+   * check has_storage() and storage_initialized().
+   */
```
Comment (ezyang) [inline c10/core/TensorImpl.h:1577]: If this is an internal helper, why is it public?

Comment (dagitses) [reply c10/core/TensorImpl.h:1577]: Tensor::data_ptr. We can make it private and friend TensorBase if we want.

File: aten/src/ATen/templates/TensorMethods.cpp [hunk 2]
```diff
@@ -1,21 +1,36 @@
 #include <c10/core/Scalar.h>
 #include <ATen/core/TensorBody.h>
 
+#include <string_view>
+
 namespace at {
 
+// Verifies the requested type is the same as the Tensor's type.
+void check_type(const TensorBase& tensor, ScalarType type, std::string_view type_name) {
```
[the remark this answers is not in the document]
Comment (dagitses) [reply aten/src/ATen/templates/TensorMethods.cpp:9]: Yes, but with string_view we know we're checking the length only once.

File: c10/core/TensorImpl.h (hunk 1 above)
Comment (ezyang) [reply c10/core/TensorImpl.h:1545]: OK. I guess it would have been better if we called this fast or unsafe, but nbd.

**Review by ezyang — APPROVED** [review]

File: c10/core/TensorImpl.h (hunk 1 above)
Comment (dagitses) [reply c10/core/TensorImpl.h:1543]: it's trivial to rename, that can be a separate change altogether.

**Pull request closed without merging**